fed/barriers.py

```diff
@@ -230,36 +235,52 @@ async def send(
             ' credentials.'
         )
         dest_addr = self._cluster[dest_party]['address']
-        dest_party_grpc_metadata = dict(
-            self._cluster[dest_party].get('grpc_metadata', {})
-        )
-        global_grpc_metadata = (
-            dict(self._grpc_metadata) if self._grpc_metadata is not None else {}
-        )
-        # merge grpc metadata
-        grpc_metadata = {**global_grpc_metadata, **dest_party_grpc_metadata}
+        dest_party_grpc_config = self.setup_dest_party_grpc_config()
+        
         try:
             response = await send_data_grpc(
                 dest=dest_addr,
                 data=data,
                 upstream_seq_id=upstream_seq_id,
                 downstream_seq_id=downstream_seq_id,
-                metadata=grpc_metadata,
+                metadata=dest_party_grpc_config["grpc_metadata"],
                 tls_config=self._tls_config,
                 retry_policy=self.retry_policy,
+                grpc_options=dest_party_grpc_config["grpc_options"]
             )
         except Exception as e:
             logger.error(f'Failed to {send_log_msg}, error: {e}')
             return False
         logger.debug(f"Succeeded to send {send_log_msg}. Response is {response}")
         return True  # True indicates it's sent successfully.
 
+    def setup_dest_party_grpc_config(self, dest_party):
+        dest_party_grpc_config = {}
+        global_grpc_metadata = (
+            dict(self._grpc_metadata) if self._grpc_metadata is not None else {}
+        )
+        dest_party_grpc_metadata = dict(
+            self._cluster[dest_party].get('grpc_metadata', {})
+        )
+        # merge grpc metadata
+        dest_party_grpc_config['grpc_metadata'] = {**global_grpc_metadata, **dest_party_grpc_metadata}
+
+        global_grpc_options = get_grpc_options(self.retry_policy)
+        dest_party_grpc_options = dict(
+            self._cluster[dest_party].get('grpc_options', [])
+        )
+        dest_party_grpc_config['grpc_options'] = {**global_grpc_options, **dest_party_grpc_options}
```

Comment:
global_grpc_options is a list?

Reply:
Either is OK. 
* grpc repo's [example](https://github.com/grpc/grpc/blob/master/examples/python/helloworld/greeter_client_with_options.py) use list of tuples
* grpc [document](https://grpc.github.io/grpc/node/grpc.Channel.html#:~:text=A%20map%20of%20channel%20options%20that%20will%20be%20passed%20to%20the%20core.%20The%20available%20options%20are%20listed%20in%20this%20document.) say it's a map

I've changed to a dict to follow the above.